codegen/src/lut.rs

```diff
@@ -0,0 +1,470 @@
+                }
+
+                impl IntoLinear<f32, u8> for #fn_type {
+                    #[inline]
+                    fn into_linear(encoded: u8) -> f32 {
+                        #table_ident[encoded as usize] as f32
+                    }
+                }
+            }
+        },
+    );
+
+    quote! {
+        #(#tables)*
+    }
+}
+
+fn build_u16_to_f64_lut(entries: &[LutEntryU16]) -> TokenStream {
+    let tables = entries.iter().map(
+        |LutEntryU16 {
+             fn_type,
+             fn_type_uppercase,
+             into_linear,
+             ..
+         }| {
+            let table = (0..=u16::MAX).map(|i| into_linear((i as f64) / 65535.0));
+            let table_ident = format_ident!("{fn_type_uppercase}_U16_TO_F64");
+            quote! {
+                #[cfg(feature = "gamma_lut_u16")]
+                static #table_ident: [f64; 65536] = [
+                    #(#table),*
+                ];
+
+                #[cfg(feature = "gamma_lut_u16")]
+                impl IntoLinear<f64, u16> for #fn_type {
+                    #[inline]
+                    fn into_linear(encoded: u16) -> f64 {
+                        #table_ident[encoded as usize]
+                    }
+                }
+
+                #[cfg(feature = "gamma_lut_u16")]
+                impl IntoLinear<f32, u16> for #fn_type {
+                    #[inline]
+                    fn into_linear(encoded: u16) -> f32 {
+                        #table_ident[encoded as usize] as f32
+                    }
+                }
+            }
+        },
+    );
+
+    quote! {
+        #(#tables)*
+    }
+}
+
+fn integrate_linear(
+    (start_x, start_t): (f64, f64),
+    (end_x, end_t): (f64, f64),
+    linear_scale: f64,
+    exp_scale: f64,
+) -> (f64, f64) {
+    let antiderive_y = |x: f64| 0.5 * linear_scale * exp_scale * x * x;
+    let antiderive_ty =
+        |x: f64, t: f64| 0.5 * linear_scale * exp_scale * x * x * (t - exp_scale * x / 3.0);
+
+    (
+        antiderive_y(end_x) - antiderive_y(start_x),
+        antiderive_ty(end_x, end_t) - antiderive_ty(start_x, start_t),
+    )
+}
+
+fn integrate_exponential(
+    (start_x, start_t): (f64, f64),
+    (end_x, end_t): (f64, f64),
+    alpha: f64,
+    gamma: f64,
+    exp_scale: f64,
+) -> (f64, f64) {
+    let antiderive_y = |x: f64, t: f64| {
+        alpha * gamma * exp_scale * x * x.powf(gamma.recip()) / (1.0 + gamma) + (1.0 - alpha) * t
+    };
+    let antiderive_ty = |x: f64, t: f64| {
+        alpha
+            * gamma
+            * exp_scale
+            * x
+            * x.powf(gamma.recip())
+            * (t - gamma * exp_scale * x / (1.0 + 2.0 * gamma))
+            / (1.0 + gamma)
+            + 0.5 * (1.0 - alpha) * t * t
+    };
+
+    (
+        antiderive_y(end_x, end_t) - antiderive_y(start_x, start_t),
+        antiderive_ty(end_x, end_t) - antiderive_ty(start_x, start_t),
+    )
+}
+
+fn build_f32_to_u8_lut(entries: &[LutEntryU8]) -> TokenStream {
```

Comment:
I would appreciate a comment that links to https://gist.github.com/2203834 for credit and explains any differences from it. Particularly the integration. It should be possible for someone who hasn't been part of the discussion to understand and bug fix it, with aid from the linked reference.